### ISSUE TITLE
sftpCommand sudo doesn't work

### ISSUE THREAD

**Issue by prog1407** [SchoofsKelvin/vscode-sshfs#104]
Thanks a lot for the nice extension. I also really need to have this sudo feature but unfortunately didn't manage to make it work for me. When I press connect it asks for password once and then status of configuration becomes orange. I can imagine that it tries to connect and ideally should ask me for sudo password once again but it doesn't

my config

`"sshfs.configs": [ { "root": "/home/red", "host": "192.168.100.100", "port": 22, "username": "red", "password": true, "name": "homeserv", "sftpCommand": "sudo /usr/lib/openssh/sftp-server" } ]`
I'm running debian 9 on my server and win10 on a client

**Comment by SchoofsKelvin** [comment]
While the extension makes use of `keyboard-interactive`, allowing (login) prompts over SSH, I'm not sure whether that system works for prompts inside programs/commands.

I might look into that, but as to make it work now: Either add the password parameter in the command (although unsafe), or change the sudoers file that makes it you don't need a password for the sftp-server command. I'm not sure how safe the latter is, though.

**Comment by DarthSpock** [comment]
It would be beneficial to sudo after a connection is established to access directories you are authorized control over but only through root ( I have a server where I access a service user home directory that can and will only be accessed via sudo)

**Comment by SchoofsKelvin** [comment]
I've added a new feature around this (3e70898) which is now live under version 1.13.0 of the extension. An example config can be found at the end of this post. While this feature seems to work fine, I still see it as a beta feature. Useful feedback (bugs, improvements, ...) would be appreciated!

Basically, when using `sftpCommand`, you can also specify `sftpSudo`. This should either be a user, e.g. `"sftpSudo": "myself"`, or `true` like `"sftpSudo": true`, which will basically call `sudo` without specifying a user, which usually defaults to `root`.

If you specify `sftpSudo` but not `sftpCommand`, it'll default to `/usr/lib/openssh/sftp-server`, which should work in most cases.

If you don't specify `sftpSudo`, but your `sftpCommand` starts with `sudo`, the extension will try to get the actual `sftpCommand` and `sftpSudo` fields from your command. As this requires me to remove flags (e.g. `-u user -S` in `sudo -u user -S /path/sftp-server`), this isn't recommended. It's also worth nothing that all sudo flags (except `-u/--user`) are ignored. Getting to run sftp-server in a sudo shell (with support for password prompt) is quite tricky, hence I can't allow a lot of customization, at least not for now.

I don't know if anyone plans on `su` instead of `sudo`. I would have to add support for the password prompt for that for it to work.

Internally, this happens when `sftpCommand` is present:
*(if only `sftpSudo` is present, `sftpCommand` first gets set to the default path)*
- Create a new ssh shell (interactive terminal)
- If `sftpSudo` is present, start a sudo shell:
  - run `sudo -S ${optionalUserFlag} bash -c "echo SUDO OK; cat | bash`
  - Prompt user for password if a message starting with `[sudo]` gets printed to stderr
  - Wait until `SUDO OK` gets printed to stdout *(after which the shell is now a sudo shell)*
- run `echo SFTP READY` and wait for this message *(skipping welcome messages)*
- run the given `sftpCommand` and wrap the stdout/stderr in a SFTPStream/SFTPWrapper

You can thus customize `sftpCommand` quite a bit, as it runs in a shell. The only issue is that it doesn't support prompting the user, e.g. when using `sudo` *(hence the whole step going around that)*. If anyone can find a way to detect when the user is prompted (VT codes? ssh event currently disabled?), I could clean this all up and make it much more easier to be customized, but as far as I know, that's not possible.

### Short version / example configs
```jsonc
    {
      "putty": "test",
      "name": "test",
      "password": true,
      "sftpCommand": "/usr/lib/openssh/sftp-server",
      "sftpSudo": true, // or a username, e.g. `"myself"`
    },
    {
      // This version works too, but isn't recommended. Most sudo flags get ignored anyway
      "putty": "test",
      "name": "test2",
      "password": true,
      "sftpCommand": "sudo -u myself /usr/lib/openssh/sftp-server", // -u flag is optional
    },
```

**Comment by DarthSpock** [comment]
@SchoofsKelvin Yeah I use `sudo su` to gain root level access with use of a password. Being able to access files on other directories outside of my `home` dir would be greatly appreciated. Workaround I have for now is to use Win32OpenSSH (or putty) to go to the directory and make a copy (or move) of the file I need to my `home` dir. If I can get to it from the workspace tree, that would be much handier.

**Comment by SchoofsKelvin** [comment]
You shouldn't need to use `su` when already using `sudo`, at least not for SFTP. Just using `sftpCommand` should be enough.

**Comment by DarthSpock** [comment]
Yeah, that's what I plan on trying out.

**Comment by DarthSpock** [comment]
So `/usr/libexec/openssh/sftp-server` is the path in Fedora. When I tried using the example above, it failed and was causing issues with the extension. It appears even inputting the correct path fails to bring up the workspace configurations after entering the setting.

**Comment by DarthSpock** [comment]
Using the non-recommended route was slightly better in that the workspace configurations appeared and it tried to connect but had a nice orange light indicator and failed to display the workspace. I attempted to disconnect but it was hung. Once again, after removing the line, I restarted Code to get it working. So this feature is buggy, at least for me.

**Comment by Kihoru** [comment]
Hey, sorry to add message here, but I can't find a way to use sftpCommand, I mean, I didn't find settings files, and there is no such field in configuration ssh fs...

**Comment by SchoofsKelvin** [comment]
The field isn't visible in the Settings UI, nor does it appear as a suggestion in `usersettings.json`. You'll have to manually add the field to your config there. This feature is a bit too much of a niche case, which is why it didn't get added to the settings UI.

**Comment by daniel347x** [comment]
Thanks for adding this feature!  Unfortunately, the instructions are not sufficiently clear for me to get it to work.  I've tried a number of combinations of things to add to the 'settings' file, but Visual Studio Code's Remote SSH connection isn't given permissions to write outside of the login user's permissions.

Any way that at some point you could write down more explicit instructions?

Thanks!

**Comment by SchoofsKelvin** [comment]
> (Note: I've already added the user to the `sudo` group and added `ALL=(ALL) ALL` permissions for the user in the `sudoers` file via `sudo visudo`.)

You might also have to enable `NOPASSWD`, as my "automatically enter sudo password" might fail to work in your case.

Opening `settings.json` and adding `"sftpSudo": "targettedUser"` should work for most systems. This assumes that running `sudo -S ${uFlag}bash -c "echo SUDO OK; cat | bash` on your server would optionally first sent a line starting with `[sudo]` which would prompt the extension to sent the password (and enter), then output "OK" and pipe stdin to bash. Not the most robust way, which might fail for your particular server setup.

**Comment by pradulmt** [comment]
Not working for a container.
This is the configuration file
    "sshfs.configs": [
        {
            "host": "10.64.219.20",
            "name": "10.64.219.20",
            "password": "somepasswdhere",
            "privateKeyPath": "/Users/pradulmt/.ssh/new_rsa",
            "sftpCommand": "/usr/lib/openssh/sftp-server",
            "sftpSudo": "pthattio",
            "username": "pthattio"
        }

If I run **sudo -S ${uFlag}bash -c "echo SUDO OK; cat | bash"** on my container, the output is **SUDO OK**, **with out asking for a password**. Any hack i can do here?

**Comment by vgorloff** [comment]
In case you are connecting to EC2  Amazon Linux AMI, then below configuration should work:

```json
"sshfs.configs": [
{
  "name": "your-config-name",
  "host": "your-host-name.compute.amazonaws.com",
  "username": "ec2-user",
  "privateKeyPath": "/path/to/ssh-key-in-ED25519-format.pem",
  "port": 22,
  "debug": false,
  "sftpSudo": true,
  "sftpCommand": "/usr/libexec/openssh/sftp-server"
}
]
```

**Comment by rdsanchezdu** [comment]
> @SchoofsKelvin Yeah I use `sudo su` to gain root level access with use of a password. Being able to access files on other directories outside of my `home` dir would be greatly appreciated. Workaround I have for now is to use Win32OpenSSH (or putty) to go to the directory and make a copy (or move) of the file I need to my `home` dir. If I can get to it from the workspace tree, that would be much handier.
> 
> I wouldn't do anything for `su` as that's an insecure method of elevating from primarily an insider threat point of view. Very difficult to track the activities of devs/sysadmins using that method.

Hi @WSLUser @SchoofsKelvin  in my work it is necessary to first log in as normal user and then run 'sudo su - myusername' I don't have much knowledge about I'm learning still,  but it would be greate if I could use this extension because I have to run some scripts and edit files all day 🥲 could you give me some help please?
